### ISSUE TITLE
Parallel-test locking, OR multiprocessing-based backend or fork?

### ISSUE THREAD

**Issue by edbrannin** [pytest-dev/pytest-xdist#115]
TL;DR: Any way to run tests in parallel but let them coordinate with locks/semaphores?

(If this should have gone to a mailing list or something, please do let me know.)

I'd _really_ like to use pytest for some live-system integration tests, but I'd also really like to run tests in parallel -- with a little bit of locking for one unfortunately-half-serial part of the application.

For example (pseudocode):
```
user_locks = collections.defaultdict(multiprocessing.Lock)
def test_1():
    # do setup stuff
    with user_locks["test_user"]:
        # Write to test_user's work-directory and kick off a slow process that reads those files.

def test_2():
    # do setup stuff
    with user_locks["test_user"]:
        # This CANNOT happen at the same time as test_1's with-lock block.

def test_3():
    # do setup stuff
    with user_locks["other_test_user"]:
        # This CAN happen concurrently with test_1 or test_2
```

So far, the closest I've come to this with another framework was using the `mp` module in nose2, which uses `multiprocessing`.  That seems to work appropriately with `multiprocessing.Lock`, but it doesn't have that special pytest flavor.

So, here are my options so far:
1. Maybe there's a way to do this with `pytest-xdist` that I didn't notice?
2. Perhaps it's feasible to add a `multiprocessing` back-end to `pytest-xdist`?
3. Perhaps it wouldn't be too hard to make my own `multiprocessing`-based `pytest` plugin.
4. I could always bite the bullet and use `nose2`.

Which option seems more reasonable to you?  If 2, please consider this as a request for it.

**Comment by nicoddemus** [comment]
Hi @edbrannin,

It is OK to post questions here, don't worry!

The short answer to your question is that no, it is not possible to use locks and semaphores in the current implementation of `pytest-xdist`. `pytest-xdist` uses `execnet` as a backend, and AFAIK it doesn't have any synchronization mechanisms.

But if all you want is some kind of lock, you can try to implement something for your test suite based on a [file lock](https://pypi.python.org/pypi/filelock/). I've done it in the past and it works well.

As for using `multiprocessing` as a backend in pytest-xdist, that's not possible in the current design. However, I feel it would be possible to write a new plugin using the current pytest-xdist codebase as starting point which used `multiprocessing` as backend, specially if you just want to use multiple cores in the same computer and drop some features like multi-python version testing support, remote testing, rsync, etc. If you keep the same architecture and use the same general design (master spawns workers, workers collect items, master coordinates test execution in workers), you should be fine. 

Take a look at our [OVERVIEW](https://github.com/pytest-dev/pytest-xdist/blob/master/OVERVIEW.md) for a more detailed description.

**Comment by edbrannin** [comment]
Hi @nicoddemus,

Thanks for your reply!  I'll compare forking pytest-xdist (because you're right, I really just want to run several IO-blocking tests concurrently; multiple cores are just a plus) vs. using the `filelock` module.

**Comment by nicoddemus** [comment]
Sure anytime.

Please post a summary of your findings here so others can benefit. 👍 

I will close this for now, feel free to follow up with more questions though.

**Comment by RonnyPfannschmidt** [comment]
note that i hope to make xdist independent of execnet in future, as a multiprocessing queue can do bascially the same job as a execnet chanesl in the context of distribtued/parallel testing

unfortunately the timeframe is rather tricky
